src/nodes.coffee

```diff
@@ -1069,6 +1084,106 @@ exports.Arr = class Arr extends Base
 
 #### Class
 
+# A node representing a Javascript class.
+# Initialize a **JsClass** with its name, an optional superclass, and a
+# list of methods.
+exports.JsClass = class JsClass extends Base
+  constructor: (@name, @parent, @methods = []) ->
+    @boundMethods = []
+
+  isStatement: YES
+
+  setConstructor: (source, constructor) ->
+    source.error 'cannot define more than one constructor in a class' if @ctor
+
+    if constructor instanceof Code
+      @ctor = constructor
+    else
+      @generatedCtor = true
+
+      apply = new Value constructor, [new Access new PropertyName 'apply']
+      @ctor = new Code
+      @ctor.body.push new Literal 'super(...arguments)' if @parent
+      @ctor.body.push new Call apply, [new ThisLiteral, new IdentifierLiteral 'arguments']
+      @ctor.body.makeReturn()
+
+    @ctor.ctor = true
+    @ctor.noReturn = true
+    @ctor.isDerivedCtor = @parent?
+
+    name = new PropertyName 'constructor'
+    @addMethod name, @ctor
+
+  addMethod: (name, method) ->
+    method.name = name
+    method.isMethod = true
+
+    if method.bound and not method.static
+      method.bound = false
+      @boundMethods.push name
+
+    # Override `method`'s `compileToFragments` to append to the class's method fragments
+    # This allows `method.compileToFragments` to be called with the correct scope whilst still
+    # being generated in the class body.
+    compileToFragments = method.compileToFragments
+    method.compileToFragments = (o) =>
```

Comment:
This is probably the worst hack in the PR. It's being used to allow methods to be compiled with the correct scope with respect to the class body, whilst also being hoisted into the `JsClass`. This is needed for the following to work:

```coffee
class A
  a = 1
  foo: ->
    a = 2 # changes `a` from the class scope
    b = 2 # declares a new variable

  b = 1
  bar: ->
    a = 2 # changes a from the class scope
    b = 2 # changes b from the class scope
```

See [the test](https://github.com/connec/coffeescript/blob/55d38af18822a3544a52d17bbd54dc81a9b8836c/test/classes.coffee#L409).